tests/shared/src/test/scala/org/typelevel/idna4s/tests/bootstring/DelimiterTests.scala

```diff
@@ -0,0 +1,154 @@
+
+        if (cp.isSurrogate) {
+          assert(Delimiter.fromInt(i).isLeft)
+          assert(Delimiter.fromCodePoint(cp).isLeft)
+        } else {
+          assert(Delimiter.fromInt(i).isRight)
+          assert(Delimiter.fromCodePoint(cp).isRight)
+        }
+
+        loop(i + 1)
+      }
+
+    loop(0)
+  }
+
+  test("Delimiter round trip") {
+
+    @tailrec
+    def loop(i: Int): Unit =
+      if (i > Character.MAX_CODE_POINT) {
+        ()
+      } else if (i === Character.MIN_SURROGATE) {
+        loop(Character.MAX_SURROGATE + 1)
+      } else {
+        val d: Delimiter = Delimiter.unsafeFromInt(i)
+        assertEquals(d.codePointInt, i)
+
+        d.codePoint.toChars match {
+          case x :: Nil =>
+            assertEquals(Delimiter.unsafeFromChar(x).codePoint, d.codePoint)
+          case x :: y :: Nil =>
+            assertEquals(Delimiter.unsafeFromInt(Character.toCodePoint(x, y)), d)
+          case _ =>
+            fail(s"Delimiter ${i} did not yield 1 or 2 chars.")
+        }
+
+        loop(i + 1)
+      }
+
+    loop(0)
+  }
+
+  test(
+    "Delimiter construction for chars should succeed for all chars which are not surrogates") {
+
+    @tailrec
+    def loop(i: Int): Unit =
+      if (i >= Char.MaxValue.toInt) {
+        ()
+      } else {
+        val cp: CodePoint = CodePoint.fromChar(i.toChar)
+        if (cp.isSurrogate) {
+          assert(Delimiter.fromChar(i.toChar).isLeft)
+        } else {
+          assert(Delimiter.fromChar(i.toChar).isRight)
+        }
+
+        loop(i + 1)
+      }
+
+    loop(0)
+  }
+
+  test("Delimiter.fromString should parse valid code points which are not surrogates") {
+
+    @tailrec
+    def loop(i: Int): Unit =
+      if (i > Character.MAX_CODE_POINT) {
+        ()
+      } else if (i === Character.MIN_SURROGATE) {
+        loop(Character.MAX_SURROGATE + 1)
+      } else {
+        assertEquals(
+          Delimiter.fromString(new String(Character.toChars(i))),
+          Right(Delimiter.unsafeFromInt(i)))
+
+        loop(i + 1)
+      }
+
+    loop(0)
+  }
+
+  test("Delimiter.fromString fails for non integral strings") {
+    assert(CodePoint.fromString("DERP").isLeft)
+  }
+
+  property("Delimiter.fromInt should fail for non code points")(
+    forAll(genNonCodePoint)(i => Prop(Delimiter.fromInt(i).isLeft) :| "Delimiter.fromInt fails")
+  )
+
+  test("Literal syntax for valid Delimiter should compile") {
+    assertEquals(delimiter"0", Delimiter.unsafeFromChar('0'))
+    assertEquals(delimiter"-", Delimiter.unsafeFromChar('-'))
+
+    // Multi char code point
+    assertEquals(delimiter"𐀀", Delimiter.unsafeFromCodePoint(codePoint"0x10000"))
+  }
+
+  test("Literal syntax for invalid literals should not compile") {
+    compileErrors("""delimiter"DERP"""").contains(
+      "Given value is not a valid non-negative integral value")
+    compileErrors("""delimiter"F"""").contains(
+      "Given value is not a valid non-negative integral value")
```

Comment:
Dumb question, why is `F` not a valid delimiter?

Reply:
It is. That's a typo. Why did that pass the tests?

Reply:
omg, I'm not asserting there. one sec